components/fxa-client/src/internal/auth_state.rs

```diff
@@ -0,0 +1,213 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::collections::{HashMap, HashSet};
+
+use crate::{
+    internal::{
+        commands, device,
+        oauth::{AccessTokenInfo, RefreshToken},
+        profile::Profile,
+        state_persistence::state_to_json,
+        CachedResponse, Config, OAuthFlow, PersistedState,
+    },
+    Result, ScopedKey,
+};
+
+/// Stores the current state of the FxA client
+///
+/// In order to make reasoning about state changes easier, all fields are private, and all state
+/// changes are confined to this module.
+pub struct AuthState {
+    /// State that's persisted to disk, which is the majority of the auth state
+    persisted_state: PersistedState,
+    /// In-progress OAuth flows
+    flow_store: HashMap<String, OAuthFlow>,
+}
+
+impl AuthState {
+    pub(crate) fn new(persisted_state: PersistedState) -> Self {
```

Comment:
Should things be `pub` or `pub(crate)`?  My current feeling is that `pub` is simpler and things still won't get exposed in the public API as long as the module is private and we don't `pub use` the type in the crate root.  Definitely open to opinions on this one though.